src/data/mod.rs

```diff
@@ -0,0 +1,249 @@
+mod cell_extractors;
+#[cfg(feature = "python")]
+mod python;
+mod rust;
+use chrono::{Duration, NaiveDate, NaiveDateTime};
+#[cfg(feature = "python")]
+pub(crate) use python::*;
+
+use calamine::{CellType, Data as CalData, DataRef as CalDataRef, DataType, Range};
+
+use crate::{
+    data::rust::{
+        create_boolean_vec, create_date_vec, create_datetime_vec, create_duration_vec,
+        create_float_vec, create_int_vec, create_string_vec,
+    },
+    error::{FastExcelErrorKind, FastExcelResult},
+    types::{
+        dtype::{DType, DTypeCoercion, get_dtype_for_column},
+        excelsheet::column_info::ColumnInfo,
+    },
+};
+
+#[derive(Debug)]
+pub(crate) enum ExcelSheetData<'r> {
+    Owned(Range<CalData>),
+    Ref(Range<CalDataRef<'r>>),
+}
+
+impl ExcelSheetData<'_> {
+    pub(crate) fn width(&self) -> usize {
+        match self {
+            ExcelSheetData::Owned(range) => range.width(),
+            ExcelSheetData::Ref(range) => range.width(),
+        }
+    }
+
+    pub(crate) fn height(&self) -> usize {
+        match self {
+            ExcelSheetData::Owned(range) => range.height(),
+            ExcelSheetData::Ref(range) => range.height(),
+        }
+    }
+
+    pub(super) fn get_as_string(&self, pos: (usize, usize)) -> Option<String> {
+        match self {
+            ExcelSheetData::Owned(range) => range.get(pos).and_then(|data| data.as_string()),
+            ExcelSheetData::Ref(range) => range.get(pos).and_then(|data| data.as_string()),
+        }
+    }
+
+    pub(crate) fn dtype_for_column(
+        &self,
+        start_row: usize,
+        end_row: usize,
+        col: usize,
+        dtype_coercion: &DTypeCoercion,
+    ) -> FastExcelResult<DType> {
+        match self {
+            ExcelSheetData::Owned(data) => {
+                get_dtype_for_column(data, start_row, end_row, col, dtype_coercion)
+            }
+            ExcelSheetData::Ref(data) => {
+                get_dtype_for_column(data, start_row, end_row, col, dtype_coercion)
+            }
+        }
+    }
+}
+
+impl From<Range<CalData>> for ExcelSheetData<'_> {
+    fn from(range: Range<CalData>) -> Self {
+        Self::Owned(range)
+    }
+}
+
+impl<'a> From<Range<CalDataRef<'a>>> for ExcelSheetData<'a> {
+    fn from(range: Range<CalDataRef<'a>>) -> Self {
+        Self::Ref(range)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum FastExcelSeries {
```

Comment:
Add doc
I would also impl `len()` and `is_empty` + `IntoIterator` for each variant